llvm/lib/Transforms/Yk/SplitBlocksAfterCalls.cpp

```diff
@@ -0,0 +1,115 @@
+//===- SplitBlocksAfterCalls.cpp -===//
+//
+// Makes function calls effectively terminators by splitting blocks after each
+// call. This ensures that there can only be at most one call per block. This
+// is used in order to detect recursion and external function calls within a
+// trace.
+
+#include "llvm/Transforms/Yk/SplitBlocksAfterCalls.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Transforms/Yk/LivenessAnalysis.h"
+
+#include <map>
+
+#define DEBUG_TYPE "yk-splitblocksaftercalls"
+
+using namespace llvm;
+
+namespace llvm {
+void initializeYkSplitBlocksAfterCallsPass(PassRegistry &);
+} // namespace llvm
+
+namespace {
+
+class YkSplitBlocksAfterCalls : public ModulePass {
+public:
+  static char ID;
+  YkSplitBlocksAfterCalls() : ModulePass(ID) {
+    initializeYkSplitBlocksAfterCallsPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    LLVMContext &Context = M.getContext();
+
+    for (Function &F : M) {
+      if (F.empty()) // skip declarations.
+        continue;
+      std::vector<BasicBlock *> Todo;
+      std::set<BasicBlock *> Seen;
+      BasicBlock &Entry = F.getEntryBlock();
+
+      Instruction *T = Entry.getTerminator();
+      for (size_t I = 0; I < T->getNumSuccessors(); I++) {
+        Todo.push_back(T->getSuccessor(I));
+      }
+      Seen.insert(&Entry);
+      while (!Todo.empty()) {
+        BasicBlock *Next = Todo.back();
+        Todo.pop_back();
+        if (Seen.count(Next) > 0) {
+          continue;
+        }
+        Seen.insert(Next);
+
+        for (Instruction &I : *Next) {
+          if (I.isDebugOrPseudoInst()) {
+            continue;
+          }
+          if (isa<CallInst>(I)) {
+            CallInst *CI = cast<CallInst>(&I);
+            if (CI->isInlineAsm()) {
+              // Inline assembler is by definition inlined so we don't need to
```

Comment:
Would inline asm containing a call cause problems?

Reply:
Good question. I hadn't thought about this. I believe it would. So we probably have to treat inline asm the same way we treat intrinsics: split the block and then determine from the trace whether it was inlined. I will leave a FIXME and then change it together with fixing intrinsics.

Reply:
Is there a way to detect a call inside inline asm? A crash would be preferred.

Reply:
On second thought. Could the inline asm contain a jump to somwhere which then does a call? I can't rule that out, so I'll rather be cautious and remove this bit entirely, and split blocks after inlineasm statements too.

Reply:
Isn't a jump morally equivalent to a call, in the sense that it splits a block? But I'm fine doing the most conservative thing and optimising later.

Reply:
I guess that's true. So in theory if we check the inline asm for calls and jumps we should be able to determine if it needs to be split at compile time. But I agree, let's make that a future optimisation.

Reply:
Checking the inline asm might turn out to be quite challenging as there's lots of ways one can encode jumps that aren't, at first, obvious. My guess is also that it'll turn out to be an optimisation with very small effects. So long as we're correct, and ideally that we leave behind a comment so that our future selves can remember what the "best" version might be, that will be fine.

Reply:
[5af6114](https://github.com/ykjit/ykllvm/pull/108/commits/5af611433ea06b43eb9f53d9b9ce1a034e7c1df3)